src/builder-manifest.c

```diff
@@ -3756,58 +3756,65 @@ builder_manifest_show_deps (BuilderManifest *self,
                               gboolean opt_yes,
                               GError         **error)
-{
-  g_autofree char *ref = NULL;
-  g_autofree char *commit = NULL;
-  g_autoptr(GPtrArray) args = NULL;
-  gboolean installed;
-
-  if (version == NULL)
-    version = builder_manifest_get_runtime_version (self);
-
-  ref = flatpak_build_untyped_ref (runtime,
-                                   version,
-                                   builder_context_get_arch (context));
 
-  commit = flatpak_info (opt_user, opt_installation, "--show-commit", ref, NULL);
-  installed = (commit != NULL);
-
-  args = g_ptr_array_new_with_free_func (g_free);
-  g_ptr_array_add (args, g_strdup ("flatpak"));
-  add_installation_args (args, opt_user, opt_installation);
-  if (installed)
-    {
-      g_ptr_array_add (args, g_strdup ("update"));
-      g_ptr_array_add (args, g_strdup ("--subpath="));
-    }
-  else
-    {
-      g_ptr_array_add (args, g_strdup ("install"));
-      g_ptr_array_add (args, g_strdup (remote));
-    }
-
-  g_ptr_array_add (args, g_strdup ("-y"));
-  if (flatpak_version_check (1, 2, 0))
-    g_ptr_array_add (args, g_strdup ("--noninteractive"));
-
-  g_ptr_array_add (args, g_strdup (ref));
-  g_ptr_array_add (args, NULL);
-
-  if (!builder_maybe_host_spawnv (NULL, NULL, 0, error, (const char * const *)args->pdata))
-    {
-      g_autofree char *commandline = flatpak_quote_argv ((const char **)args->pdata);
-      g_prefix_error (error, "running `%s`: ", commandline);
-      return FALSE;
-    }
+{
+  const char *remote = *remotes;
+  while (remote != NULL) {
+    g_autofree char *ref = NULL;
+    g_autofree char *commit = NULL;
+    g_autoptr(GPtrArray) args = NULL;
+    gboolean installed;
+
+    if (version == NULL)
+      version = builder_manifest_get_runtime_version (self);
+
+    ref = flatpak_build_untyped_ref (runtime,
+				     version,
+				     builder_context_get_arch (context));
+
+    commit = flatpak_info (opt_user, opt_installation, "--show-commit", ref, NULL);
+    installed = (commit != NULL);
+
+    args = g_ptr_array_new_with_free_func (g_free);
+    g_ptr_array_add (args, g_strdup ("flatpak"));
+    add_installation_args (args, opt_user, opt_installation);
+    if (installed)
+      {
+	g_ptr_array_add (args, g_strdup ("update"));
+	g_ptr_array_add (args, g_strdup ("--subpath="));
+      }
+    else
+      {
+	g_ptr_array_add (args, g_strdup ("install"));
+	g_ptr_array_add (args, g_strdup (remote));
+      }
+
+    g_ptr_array_add (args, g_strdup ("-y"));
+    if (flatpak_version_check (1, 2, 0))
+      g_ptr_array_add (args, g_strdup ("--noninteractive"));
+
+    g_ptr_array_add (args, g_strdup (ref));
+    g_ptr_array_add (args, NULL);
+
+    if (!builder_maybe_host_spawnv (NULL, NULL, 0, error, (const char * const *)args->pdata))
+      {
+	g_autofree char *commandline = flatpak_quote_argv ((const char **)args->pdata);
+	g_prefix_error (error, "running `%s`: ", commandline);
```

Comment:
We need to handle errors and looping better. For instance, on an update we should never try again with another remote, because we didn't even use the remote argument. 

Then, if we get a error other than the flatpak command exited with exit code != 0 then we should not loop (i.e. if we have some internal error kind of thing). 

But, if we get an exit code != 0 on an install, we have to assume that it printed the error already. And, if we have more that one remote such errors are going to be expected which is not obvious to the user. So, in the multi-remote case we should print something like "Trying install from remote %s" or something, and then the error should only be set at the last remote failure. (The above is actually problematic in that it sets error multiple times.)

Reply:
I don't get the difference between install and update. Maybe I missed some flow, it looked initially like the check in main prevented entering these install functions if remote is not given. Will review the code again.

Yeah, I guess logging needs improving.

Reply:
The CLI syntax for install is "$remote $id", whereas for update it is just "$id" and always pulls from previous origin.

Reply:
I'm still confused. This is flatpak-builder, I can't find any explicit update command.

Reply:
Ah. So we call flatpak to install. Fair enough.

Reply:
I'll try to refactor these flows to separate internal functions , it should help help.

Reply:
Should error only be set once? You can't know beforehand which remote is supposed to have the ref.

Reply:
@alexlarsson maybe we should be more generic with the error anyway? Looked like process command line is already outputted on debug level.

Reply:
Output is now clearer. I'm still trying to research how the error handling should work.

Reply:
Eg the problem is not just that error was set by us repeatedly. Spawning the child process also sets the error so currently the code still bugs if there's 2+ first remotes that don't have the ref.

Reply:
This is not done. I haven't figured out how to disambiguate the flatpak running errors from other errors.

Reply:
Now checking G_SPAWN_EXIT_ERROR domain for the GError, this seems to work.

Reply:
Hmm, the logic I wrote is actually subtly incorrect and it's now sometimes setting error even if we succeed.

Reply:
The cleanest way I could figure out is just collecting error always but finally resetting it if the installation was successful. Is this ok?